src/modules/update/common/update-common.c

```diff
@@ -0,0 +1,488 @@
+    if (handle->timeout)
+        sol_timeout_del(handle->timeout);
+    if (handle->conn)
+        sol_http_client_connection_cancel(handle->conn);
+    free(handle);
+}
+
+static void
+free_response_members(struct sol_update_info *info)
+{
+    free((void *)info->version);
+    free((void *)info->url);
+    free((void *)info->hash);
+    free((void *)info->hash_algorithm);
+}
+
+static void
+task_check_response(void *data, const struct sol_http_client_connection *conn,
+    struct sol_http_response *http_response)
+{
+    struct sol_update_handle *handle = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    struct sol_update_info response = { };
+
+    handle->conn = NULL;
+
+    if (http_response->response_code != SOL_HTTP_STATUS_OK) {
+        SOL_WRN("Invalid response code from [%s] when checking for update: %d",
+            handle->url, http_response->response_code);
+        handle->cb_check((void *)handle->user_data, -http_response->response_code, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    if (!streq(http_response->content_type, "application/json")) {
+        SOL_WRN("Invalid content type of response: [%s] expected [application/json]",
+            http_response->content_type);
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    if (!http_response->content.used) {
+        SOL_WRN("Empty response for version check");
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    sol_json_scanner_init(&scanner, http_response->content.data, http_response->content.used);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "url"))
+            response.url = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "version"))
+            response.version = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash"))
+            response.hash = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash-algorithm"))
+            response.hash_algorithm = sol_json_token_get_unescaped_string_copy(&value);
+        else
+            SOL_WRN("Unknown response member: %.*s",
+                SOL_STR_SLICE_PRINT(sol_json_token_to_slice(&token)));
+    }
+
+    /* If some member is empty, give up */
+    if (!(response.url && response.version
+        && response.hash && response.hash_algorithm)) {
+        SOL_WRN("Malformed check response");
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        free_response_members(&response);
+        delete_handle(handle);
+        return;
+    }
+
+    handle->cb_check((void *)handle->user_data, 0, &response);
```

Comment:
need to flag somehow, otherwise calling "cancel" from handle->cb_check will crash... you know, people do these things :-D